File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        
     List<List<Integer>> ans = new ArrayList<>();
        calc(root,  ans,0);
        return ans;
    }

    private void calc(TreeNode root, List<List<Integer>> ans, int level) {
        if(root==null)return;
        if(ans.size()<=level){
            ans.add(new ArrayList<>());
        }
        if(level%2==0){
            ans.get(level).add(root.val);
        }else{
            ans.get(level).add(0,root.val);
        }
        calc(root.left,ans,level+1);
        calc(root.right,ans,level+1);
    }
}

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        q=deque()
        q.append(root)
        an=[[root.val]]
        mx=1
        while q: 
            ll=len(q)
            larr=[]
            for _ in range(ll):
                cur=q.popleft() 
                if cur.left:
                    q.append(cur.left)
                    larr.append(cur.left.val)
                if cur.right:
                    q.append(cur.right)
                    larr.append(cur.right.val)
            if len(larr)>0:
                if mx%2==0:
                    an.append(larr)
                else:
                    an.append(larr[::-1])
                    
            mx+=1
        return an
        